File: 1478-maximum-number-of-events-that-can-be-attended/1478-maximum-number-of-events-that-can-be-attended.py
class Solution:
    def maxEvents(self, events: List[List[int]]) -> int:

        events.sort()
        days = 0
        
        for event in events:
            days = max(days, event[1])

        i = 0
        n = len(events)
        min_heap = []
        result = 0

        for day in range(1, days + 1):

            while i < n and events[i][0] <= day:
                heapq.heappush(min_heap, events[i][1])
                i += 1

            while min_heap and min_heap[0] < day:
                heapq.heappop(min_heap)

            if min_heap:
                heapq.heappop(min_heap)
                result += 1

        
        return result
            

        
        

File: 1478-maximum-number-of-events-that-can-be-attended/1478-maximum-number-of-events-that-can-be-attended.go
import (
    "container/heap"
    "sort"
)

type minHeap []int

func (h minHeap) Len() int {
    return len(h)
}

func (h minHeap) Less(x, y int) bool {
    return h[x] < h[y]
}

func (h minHeap) Swap(x, y int) {
    h[x], h[y] = h[y], h[x]
}

func (h *minHeap) Push(x any) {
    *h = append(*h, x.(int))
}

func (h *minHeap) Pop() any {
    old := *h
    x := old[len(old) - 1]
    *h = old[:len(old) - 1]
    return x
}


func maxEvents(events [][]int) int {
    sort.Slice(events, func (i, j int) bool {
        return events[i][0] < events[j][0]
    })

    mHeap := &minHeap{}
    heap.Init(mHeap)

    i, n, day, result, maxDay := 0, len(events), 1, 0, 0

    for _, event := range events {
        if maxDay < event[1] {
            maxDay = event[1]
        }
    }

    for day = 1; day <= maxDay; day++ {
        for i < n && events[i][0] == day {
            heap.Push(mHeap, events[i][1])
            i++
        }

        for mHeap.Len() > 0 && (*mHeap)[0] < day {
            heap.Pop(mHeap)
        }

        if mHeap.Len() > 0 {
            heap.Pop(mHeap)
            result++
        }
    }

    return result
}